Add environment variable & CLI flag overrides for Timeout & Retry settings

laguirre-cs opened the pull request:
# Checklist:
Please check you fulfill ALL of the relevant checkboxes
- [ ] Notified docs of any potential USER-facing changes
- [x] Added short description of the change - with relevant motivation and context. 
- [ ] Branch has the ticket number in its name (along with a ticket summary) <-- woops
- [x] Commented the code, particularly in hard-to-understand areas
- [x] Added tests that prove my fix is effective or that my feature works
- [x] Ran `npm test` and it passes
- [x] Changes generate no new warnings
- [x] Changes are backward compatible with older clusters

Changes:
* Adds environment variable overrides for `got`'s Timeout & Retry values.
* Exposes above in the `cortex configure env` command
* Adds `--no-timeout`  flag that disables timeouts
* Fixes minor bug where `Using CORTEX_TOKEN from ...` message was printed multiple times
* Minor error handling update for timeouts - prints the information on how to resolve the error

Related Issue: https://cognitivescale.atlassian.net/browse/FAB-6392

laguirre-cs committed:
Add ability to set custom timeouts via environment variables

laguirre-cs committed:
Resolve edge case where multiple messages are printed about using CORTEX_TOKEN env var

jgielstra commented:
export CORTEX_TIMEOUT_LOOKUP=                               (default: 100, unit: ms)
export CORTEX_TIMEOUT_CONNECT=                              (default: 50, unit: ms)
export CORTEX_TIMEOUT_SECURE_CONNECT=                       (default: 100, unit: ms)
export CORTEX_TIMEOUT_SOCKET=                               (default: 1000, unit: ms)
export CORTEX_TIMEOUT_RESPONSE=                             (default: 2000, unit: ms)

^^ Not valid env vars ..  added '#' before (...)

eval $(cortex configure env) << Command used like this

laguirre-cs commented:
> export CORTEX_TIMEOUT_LOOKUP= (default: 100, unit: ms) export CORTEX_TIMEOUT_CONNECT= (default: 50, unit: ms) export CORTEX_TIMEOUT_SECURE_CONNECT= (default: 100, unit: ms) export CORTEX_TIMEOUT_SOCKET= (default: 1000, unit: ms) export CORTEX_TIMEOUT_RESPONSE= (default: 2000, unit: ms)
> 
> ^^ Not valid env vars .. added '#' before (...)
> 
> eval $(cortex configure env) << Command used like this

Good idea! Will do make it compatible with that command. I was also on the fence about including the `(default: 50, unit: ms)` part. I thought it would be helpful to publish that info because it's not obvious, but wasn't sure how to best format that.

laguirre-cs committed:
Add commenting to always create valid env vars & support 'eval export CORTEX_TOKEN=eyJhbGciOiJFZERTQSIsImtpZCI6InpkYVA3VktQQXEzT2FhVXV6d3laWDhOSG5EbENidXZ4ZDhBNmhXdkZBQTgifQ.eyJzdWIiOiI1NWE0YTlmYS1hNzNhLTRjYzgtYWI1ZS1iMTc2MDA0YzQ4ODIiLCJhdWQiOiJzZW5zYSIsImlzcyI6InRlY25vdHJlZS5jb20iLCJpYXQiOjE3MDI1Njg1OTUsImV4cCI6MTcwMjY1NDk5NX0.T_8rwd3vsgqNDPM3LHC5ZGpSByT_KFyteFqR3yq56fERx7EdRATEqdUisIKzZzKfE-QGbAeg1H4QE0I2k3FnCw
export CORTEX_URI=https://api.dci-dev.dev-eks.insights.ai
export CORTEX_URL=https://api.dci-dev.dev-eks.insights.ai
export CORTEX_PROJECT=laguirre-testi-69257
export CORTEX_TIMEOUT_LOOKUP=1000                           # (unit: ms)
export CORTEX_TIMEOUT_CONNECT=1000                          # (unit: ms)
export CORTEX_TIMEOUT_SECURE_CONNECT=1000                   # (unit: ms)
export CORTEX_TIMEOUT_SOCKET=1000                           # (unit: ms)
export CORTEX_TIMEOUT_RESPONSE=2000                         # (unit: ms)'

laguirre-cs committed:
Fix spacing issue

laguirre-cs committed:
Update unit tests

laguirre-cs committed:
Revert default # retries to 3 and add CORTEX_API_RETRY env var override

laguirre-cs committed:
Increment lookup and connect default timeout values

laguirre-cs committed:
Remove duplicate code between constructError() & handleError() in utils

laguirre-cs committed:
Update error handling to better report errors from timeout

laguirre-cs committed:
Fix line count

laguirre-cs committed:
Add --no-timeout flag to disable all timeouts - takes priority over env vars

laguirre-cs commented:
See this comment for an overview of the changes in this PR: https://cognitivescale.atlassian.net/browse/FAB-6392?focusedCommentId=105455

laguirre-cs committed:
Fix wording

laguirre-cs committed:
Remove hardcoded token

laguirre-cs committed:
fix linting error

yreddy-CS committed:
Merge branch 'develop' into timeout-env-overrides

laguirre-cs committed:
Fix unit tests checking for dynamic value for token in Cortex Configure Env

laguirre-cs committed:
Add additional help text the to the configure env commmand

laguirre-cs committed:
Add additional reference text as bash comments to configure env command and make test assertion easier to maintain

laguirre-cs commented:
This should be working & ready for review now (unit tests now pass in the pipeline). In addition I include some extra bash comments in the output of `cortex configure env`. Refer to the linked JIRA comment for a description of the changes
